[WIP] Persist DAG on disk

itegulov opened the pull request:
## Overview
Implements the rest parts of `BlockDagStorage`: `childMap` storage, `dataLookup` storage and `toposort` storage.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/RHOL-539

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] If this PR adds a new feature, this PR includes tests related to this feature.
- [ ] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [developer wiki](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain) for instructions.
- [ ] Your GitHub account is also an account with [Travis CI](https://travis-ci.org). Unit tests will not run on your PR unless you have an account with Travis. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

birchmd commented on block-storage/src/main/scala/coop/rchain/blockstorage/BlockMetadata.scala:
```diff
@@ -1,19 +1,54 @@
     seqNum: Int
-)
+) {
+  def toByteString: ByteString = {
```
I wonder if this method, along with `fromBytes` would be an argument for just making this class be part of the protobuf model (then we would get them for free).

itegulov replied:
I think this makes sense. It will also allow us to modify `BlockMetadata` structure seamlessly in the future.

itegulov commented:
I am almost finished with this: implemented checkpoint loading, storage for block metadata (`childMap` and `topoSort` are not stored separately and are instead built from block metadata log) and addressed @birchmd's comments about effectful operations. I will add more corner-case tests and implement `checkpoint()` function properly shortly.

@birchmd PTAL

Primetalk commented on block-storage/src/test/scala/coop/rchain/blockstorage/BlockDagStorageTest.scala:
```diff
@@ -43,116 +70,310 @@ class BlockDagFileStorageTest extends BlockDagStorageTest {
+    latestMessages shouldBe blockElements.map(b => b.sender      -> BlockMetadata.fromBlock(b)).toMap
+
+    def normalize(topoSort: Vector[Vector[BlockHash]]): Vector[Vector[BlockHash]] =
+      if (topoSort.size == 1 && topoSort.head.isEmpty)
```
`topoSort.filter(_.nonEmpty)`?

itegulov commented on block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagStorage.scala:
```diff
@@ -22,7 +22,7 @@ trait BlockDagRepresentation[F[_]] {
   def contains(blockHash: BlockHash): F[Boolean]
   def topoSort(startBlockNumber: Long): F[Vector[Vector[BlockHash]]]
-  def topoSortTail(tailLength: Long): F[Vector[Vector[BlockHash]]]
+  def topoSortTail(tailLength: Int): F[Vector[Vector[BlockHash]]]
```
`tailLength` basically represents the length of the resulting vector. Since we can't create vectors longer than 2^31-1, it does not make sense to allow to specify a `Long` length for it. On the other hand, `startBlockNumber` just represents the offset which could potentially be larger than 2^31-1, but it is okay since the entire data structure won't be stored in memory.

Primetalk commented on block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala:
```diff
@@ -382,16 +374,12 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: BlockStore] private
-          }
-      _ <- lock.release
+      _                             <- lock.acquire
+      latestMessagesLogOutputStream <- latestMessagesLogOutputStreamRef.get
```
INFO: 
```scala
_ <- latestMessagesLogOutputStreamRef.get.map(_.close())
```

itegulov commented:
@Primetalk thanks for your review, it was very helpful!
